pkg/agent/cniserver/pod_configuration_test.go

```diff
@@ -0,0 +1,464 @@
+		ofPortAssigned     bool
+		podIfaceIsCached   bool
+		installFlow        bool
+		flowInstalled      bool
+		installOpenFlowErr error
+		expErr             string
+		expEvent           string
+	}{
+		{
+			name:             "updated Port is not in interface store",
+			podIfaceIsCached: false,
+			installFlow:      false,
+		}, {
+			name:             "OpenFlow port is not assigned",
+			podIfaceIsCached: true,
+			ofPortAssigned:   false,
+			installFlow:      false,
+			expErr:           "pod's OpenFlow port is not ready yet",
+			expEvent:         "Warning NetworkNotReady Pod network forwarding rules not installed",
+		}, {
+			name:               "failed to install OpenFlow entries for updated Port",
+			podIfaceIsCached:   true,
+			ofPortAssigned:     true,
+			installFlow:        true,
+			installOpenFlowErr: fmt.Errorf("failure to install flow"),
+			expErr:             "failed to add Openflow entries for OVS port test: failure to install flow",
+			expEvent:           "Warning NetworkNotReady Pod network forwarding rules not installed",
+		}, {
+			name:               "succeeded",
+			podIfaceIsCached:   true,
+			ofPortAssigned:     true,
+			installFlow:        true,
+			installOpenFlowErr: nil,
+			expEvent:           "Normal NetworkReady Installed Pod network forwarding rules",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+
+			waiter := newAsyncWaiter(testPodNameA, podInfraContainerID, stopCh)
+
+			testClients := newMockClients(controller, nodeName, pod)
+			testClients.startInformers(stopCh)
+			fakeOFClient := testClients.ofClient
+
+			configurator := newTestPodConfigurator(testClients, waiter)
+
+			flowInstalled := false
+
+			ifConfig := interfacestore.InterfaceConfig{
+				InterfaceName: podIfName,
+				ContainerInterfaceConfig: &interfacestore.ContainerInterfaceConfig{
+					PodNamespace: testPodNamespace,
+					PodName:      testPodNameA,
+					ContainerID:  podInfraContainerID,
+				},
+				OVSPortConfig: &interfacestore.OVSPortConfig{
+					PortUUID: "test-port-uuid",
+				},
+				IPs: podIPs,
+				MAC: podMac,
+			}
+
+			if tc.ofPortAssigned {
+				ifConfig.OVSPortConfig.OFPort = int32(1)
+			}
+
+			if tc.podIfaceIsCached {
+				configurator.ifaceStore.AddInterface(&ifConfig)
+			}
+
+			if tc.installFlow {
+				fakeOFClient.EXPECT().InstallPodFlows(podIfName, podIPs, podMac, portStatusMsg.Desc.PortNo, uint16(0), nil).Times(1).Return(tc.installOpenFlowErr)
+				if tc.installOpenFlowErr == nil {
+					flowInstalled = true
+				}
+			}
+
+			err := configurator.updateUnreadyPod(podIfName)
+			if tc.expErr == "" {
+				require.NoError(t, err)
+			} else {
+				require.EqualError(t, err, tc.expErr)
+			}
+
+			if flowInstalled {
+				assert.True(t, waiter.waitUntil(5*time.Second))
+			}
+
+			if tc.expEvent != "" {
+				actEvt := <-testClients.recorder.Events
+				require.Equal(t, tc.expEvent, actEvt)
+			}
```

Comment:
```suggestion
			var gotEvent string
			select {
			case gotEvent = <-testClients.recorder.Events:
			default:
			}
			require.Equal(t, tc.expEvent, gotEvent)
```
This is why it didn't catch the case that misses event, and it wouldn't wait forever when code is wrong.

Reply:
updated.